### ISSUE TITLE
circle loss

### ISSUE THREAD

**Issue by manongXT** [hrlblab/CircleNet#14]
Hello, thank you for your code sharing. I encountered the following problems when training my dataset: first, the training loss seems to be the same as that of centernet, with no circle loss. Second, Why is cIOU not reflected in loss.
Hope to get your answer, thank you.

**Comment by EthanHNguyen** [comment]
Hello @manongXT. Thanks for your interest in our project. 

Can you please share the command you use to train CircleNet? The main difference is that the circle loss contains a loss function to optimize the radius of the circle.

On cIOU, you can also ask why IOU is not used in loss function. cIOU is primarily as an evaluation metric. However, it may be interesting to test cIOU in a loss function.

**Comment by manongXT** [comment]
Thank you for your reply.
My training command is'python main.py --task circledet --exp_id CircleNet_DLA_Reproduce --arch dla_34 --batch_size 4 --master_batch 4 --lr 2.5e-4 --load_model /home/huangzhitian/CircleNet/models/ctdet_coco_dla_2x.pth'.Is the radius code of the optimized circle in the training loss in this file(../CircleNet/src/lib/trains/circledet.py):
`if opt.wh_weight > 0:
        if opt.dense_wh:
          mask_weight = batch['dense_wh_mask'].sum() + 1e-4
          cl_loss += (
            self.crit_wh(output['wh'] * batch['dense_wh_mask'],
            batch['dense_wh'] * batch['dense_wh_mask']) /
            mask_weight) / opt.num_stacks
        elif opt.cat_spec_wh:
          cl_loss += self.crit_wh(
            output['wh'], batch['cat_spec_mask'],
            batch['ind'], batch['cat_spec_wh']) / opt.num_stacks
        else:
          cl_loss += self.crit_reg(
            output['cl'], batch['reg_mask'],
            batch['ind'], batch['cl']) / opt.num_stacks`
The last else is to calculate the loss of circle radius?How does the code reflect the radius loss formula of the circle in the paper？

**Comment by manongXT** [comment]
Thank you for your reply.
My training command is'python main.py --task circledet --exp_id CircleNet_DLA_Reproduce --arch dla_34 --batch_size 4 --master_batch 4 --lr 2.5e-4 --load_model /home/huangzhitian/CircleNet/models/ctdet_coco_dla_2x.pth'.Is the radius code of the optimized circle in the training loss in this file(../CircleNet/src/lib/trains/circledet.py):
if opt.wh_weight > 0: if opt.dense_wh: mask_weight = batch['dense_wh_mask'].sum() + 1e-4 cl_loss += ( self.crit_wh(output['wh'] * batch['dense_wh_mask'], batch['dense_wh'] * batch['dense_wh_mask']) / mask_weight) / opt.num_stacks elif opt.cat_spec_wh: cl_loss += self.crit_wh( output['wh'], batch['cat_spec_mask'], batch['ind'], batch['cat_spec_wh']) / opt.num_stacks else: cl_loss += self.crit_reg( output['cl'], batch['reg_mask'], batch['ind'], batch['cl']) / opt.num_stacks
The last else is to calculate the loss of circle radius?How does the code reflect the radius loss formula of the circle in the paper？

**Comment by manongXT** [comment]
Thank you for your reply.
My training command is'python main.py --task circledet --exp_id CircleNet_DLA_Reproduce --arch dla_34 --batch_size 4 --master_batch 4 --lr 2.5e-4 --load_model ../CircleNet/models/ctdet_coco_dla_2x.pth'.
Is the radius code of the optimized circle in the training loss in this file(../CircleNet/src/lib/trains/circledet.py):
if opt.wh_weight > 0: if opt.dense_wh: mask_weight = batch['dense_wh_mask'].sum() + 1e-4 cl_loss += ( self.crit_wh(output['wh'] * batch['dense_wh_mask'], batch['dense_wh'] * batch['dense_wh_mask']) / mask_weight) / opt.num_stacks elif opt.cat_spec_wh: cl_loss += self.crit_wh( output['wh'], batch['cat_spec_mask'], batch['ind'], batch['cat_spec_wh']) / opt.num_stacks else: cl_loss += self.crit_reg( output['cl'], batch['reg_mask'], batch['ind'], batch['cl']) / opt.num_stacks
The last else is to calculate the loss of circle radius?How does the code reflect the radius loss formula of the circle in the paper？

**Comment by EthanHNguyen** [comment]
Yes, [that](https://github.com/hrlblab/CircleNet/blob/2e8f557c1d8c2b6b63ad462eaa1139d9bd1c2273/src/lib/trains/circledet.py#L58) is the correct line of code. The name is misleading but `wh` refers to the radius in this context. The rationale is that `wh` requires 2 degrees of freedom, and you can obtain `radius` by reducing `wh` to 1 degree of freedom. Sorry for the confusion.

Does that make sense?

**Comment by manongXT** [comment]
![image](https://user-images.githubusercontent.com/40347596/176901269-56aeb35a-8499-47c0-a40e-f7aef9d93524.png)
where the code reflect this formula?

**Comment by EthanHNguyen** [comment]
I mixed this up with another line experiment. 

That formula is reflected here: https://github.com/hrlblab/CircleNet/blob/2e8f557c1d8c2b6b63ad462eaa1139d9bd1c2273/src/lib/trains/circledet.py#L67-L69

`crit_reg` defaults to an L1 Loss as described by the formula.
